canopy_config.ml

```
@@ -21,14 +13,28 @@ let decompose_git_url url =
 let remote_uri () = fst (decompose_git_url (Key_gen.remote ()))
 let remote_branch () = snd (decompose_git_url (Key_gen.remote ()))
 
-let config () = {
-  remote_uri = remote_uri ();
-  remote_branch = remote_branch ();
-  index_page = Key_gen.index ();
-  blog_name = Key_gen.name ();
-  port = Key_gen.port ();
-  push_hook_path = Key_gen.push_hook ();
-  tls_port = Key_gen.tls_port ();
-  uuid = Key_gen.uuid ();
-  root = Key_gen.root ();
-}
+let index_page cache =
+  match KeyMap.find_config_opt cache [".config";"index_page"] with
+    None -> "Index"
+  | Some p -> p
+
+let blog_name cache =
+  match KeyMap.find_config_opt cache [".config";"blog_name"] with
+    None -> "Canopy"
+  | Some n -> n
+
+let uuid cache =
+  match KeyMap.find_config_opt cache [".config";"uuid"] with
+    None -> raise (Required_config "uuid")
+  | Some u -> u
```

wonder when this `raises` if the repo does not contain a uuid... upon pull for the first time, or when the uuid is actually used (the first client requests the atom feed)?  former is fine, the latter would be inconvenient.

---

At the moment this function is only called from within `Canopy_syndic.atom` this would occur whenever `update_atom` is used... so it is not limited to startup time (for instance it happens if we request the `push_hook_path`)

---

So to handle a missing `uuid`, we could:
* let the exception bubble to `Canopy_main.start` at startup time
* catch the exception and return an empty list for the atom feed, in `Canopy_syndic.atom`

How does that sound ?

---

hmm, what about: crash on startup if there's no uuid, and preserve this one uuid over the entire lifetime (since it is not meant to change anyways)... if someone insists to change their uuid, they can do so by pushing a new and restart Canopy manually... then we'll access this (potentially raising) function only once, and never deliver an empty atom feed :)

---

ah yes, sounds good! 👍